esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -178,6 +178,47 @@ void CldapenvironmentEx::createLDAPBaseDN(const char * baseDN, SecPermissionType
     }
 }
 
+bool CldapenvironmentEx::createUser(StringBuffer &userName, const char * prefix, const char * envName, const char * baseDN, const char * pwd, StringBuffer &notes)
+{
+#define MAX_USERNAME_LDAP_LEN 20
+    StringBuffer tmpUserName;
+    tmpUserName.appendf("%s%s", prefix, envName);
+    userName.set(tmpUserName);
+
+    Owned<ISecUser> user = secmgr->createUser(tmpUserName.str());
+    user->credentials().setPassword(pwd);
+
+    int idx = 0;
+    while (true)
+    {
+        try
+        {
+            DBGLOG("Creating user '%s' in '%s', len=%d", tmpUserName.str(), baseDN, tmpUserName.length());
+            secmgr->addUser(*user.get(), baseDN);
+            userName.set(tmpUserName.str());//success!
+            return true;
+        }
+        catch(...)
+        {
+            if (idx == 99)
+            {
+                notes.appendf("\nNon Fatal Error creating '%s'", userName.str());
+                return false;
+            }
+            if (idx == 0 && tmpUserName.length() > MAX_USERNAME_LDAP_LEN)
+                tmpUserName.setLength(MAX_USERNAME_LDAP_LEN);//LDAP limits account names to 20 characters
+            else
+            {
+                char szIndex[3];//'1' through '99'
+                itoa(++idx, szIndex, 10);
+                strcpy((char *)(tmpUserName.str() + tmpUserName.length() - strlen(szIndex)), szIndex);
```

Comment:
You are making big assumptions about StringBuffer internals here.  Especially not getting out of sync with the internal length.

I would:
```
   char szIndex[3];//'1' through '99'
   itoa(++idx, szIndex, 10);
   tmpUserName.setLength(tmpUserName.length() - strlen(szIndex));
   tmpUserName.append(szIndex);
```
or
```
   tmpUserName.setLength(tmpUserName.length() - ((idx > 9) ? 2 : 1));
   tmpUserName.append(idx);
```

Reply:
I am never changing the length of the StringBuffer .
I copy to "tmpUserName.str() + tmpUserName.length() - strlen(szIndex))" ,  for a length of "strlen(szIndex)",
so it's always (and only) the last 1 or 2 characters that get overwritten.

Reply:
The fact that you have to cast away the const says something about going against the intent. But hmmm.  It will work.  And is preserving the length as is.  I'd be curious to see what @ghalliday thought of this.

Reply:
@afishbeck @ghalliday   I went ahead and made the change

Reply:
I agree the new code is much better, and old code was a bit dubious..